### ISSUE TITLE
Add capability to specify fixture and test hooks on the test run level

### ISSUE THREAD

**Issue by benmonro** [DevExpress/testcafe#4323]
### What is your Test Scenario?

I would like to add a global/suite-level before/beforeEach after/afterEach that is NOT run client side, but rather on the node side.  We have some setup that needs to be performed in node, before a test setup is executed.  It would be nice to be able to do this in a global setup / config.

### What are you suggesting?

In https://github.com/DevExpress/testcafe/issues/3857 I was asking for what eventually became clientScripts.  But now I need the ability to have a 'before' that is global to all tests, but not executed in the client.

### What alternatives have you considered?

there are none as far as I know.
### Additional context

so in our use case, we have to establish a 'sessionId' for use with our server, which must be passed in via the page URL.

I have to somehow get the sessionId into the page url, but can't seem to do that. 

```
fixture`my page`
    .before(async ctx => {
        sessionId = await registerSession();
        ctx.sessionId = sessionId;
    })
    .page`${BASE_URL}?sessionId=${sessionId}`;
```

is there any way to do this?

**Comment by alexey-lin** [comment]
Hi @benmonro,
 
You can use [Fixture Hooks](https://devexpress.github.io/testcafe/documentation/test-api/test-code-structure.html#fixture-hooks). They are executed on the node side before the test setup.
 
As for your code snippet, you can use something like this:
```js
fixture`my page`
    .before(async ctx => {
        const sessionId = await registerSession();
        ctx.sessionId = sessionId;
    })
    .beforeEach(async t => {
        await t.navigateTo(`${BASE_URL}?sessionId=${t.fixtureCtx.sessionId}`);
    });
```

**Comment by benmonro** [comment]
Yes but I can't apply this to all fixtures.  I would like all fixtures to have this session id. Is that possible?

**Comment by miherlosev** [comment]
@benmonro 

As I understand you want to specify the `before` hook in a single place and then this hook will be executed for each `fixture`. Theoretically, it can be useful. But I cannot imagine the real use case for which it's necessary.
So, could you please describe your use case in greater detail? What is the `registerSession` function doing?

**Comment by benmonro** [comment]
@miherlosev yes the register session is something we need to use with our back end so that tests can run concurrently without stopping in each others data.  We are required to use this api and no mock http requests.  That means every single fixture will have this in it's before/each.  Would be really nice to have this done in one place rather than in each test file

**Comment by benmonro** [comment]
Thank you @miherlosev this will be a great enhancement.  Gotta give you and the whole team big props man for putting up with my schtick.  Testcafe is the stone cold best e2e tool out there. <3 Testcafe

**Comment by benmonro** [comment]
@miherlosev any update on this?  I know it's been a while, but Walmart is now looking to adopt testcafe company-wide, and this feature will be very important...

**Comment by AlexanderMoiseev** [comment]
Hello, this feature is in our TO-DO list but I can’t give you any timeframe at the moment. Please follow this thread to be informed of our progress.

**Comment by benmonro** [comment]
@AlexanderMoiseev how can we raise the priority of this task?

**Comment by Dmitry-Ostashev** [comment]
@benmonro The more users vote for this feature, the more priority it has. For now, it is not in our nearest development plans. You can submit your implementation in the meantime.

**Comment by miherlosev** [comment]
The `runner.beforeEachTest` hook will be usefull for [this](https://stackoverflow.com/questions/63635483/how-to-login-only-once-before-each-fixture-in-testcafe) scenario.

**Comment by benmonro** [comment]
Every other test framework has a global setup like this.  It would be so useful in a variety of use cases.  Checking a11y after each test in the suite for example. Enabling & configuring mock endpoints is another example.

**Comment by AndreyBelym** [comment]
As a workaround, you can create the following `setup.js` file:
```
export const fixture = (...args) => global.fixture(...args)
   .beforeEach(async t => {
      console.log('each');
   });
```

Now you can use youк own `fixture` to declare fixtures with predefined hooks in any other file:
```js
import { fixture } from './setup.js';

 

fixture `My fixture`
    .page `example.com`;
```

**Comment by benmonro** [comment]
ok @AndreyBelym is the plan still to build this into testcafe?  the above might work, but it's not as nice as a global setup supported by Testcafe...

**Comment by AndreyBelym** [comment]
There are multiple suggestions for improving the test hook subsystem. The amount of work seems quite big. Currently, we are focused on finishing our current roadmap, especially the multiple browser windows feature. We will consider hook enhancements for the next roadmap.

**Comment by miherlosev** [comment]
We decided to implement this via the global setup/teardown conception.
Presumably, it will be a pair of options: `globalSetup`, `globalTeardown`.
Each option contains a path to the file with setup actions.
The file contains the setup/teardown logic that will be applied for all tests.
*Example*

```js
testcafe chrome tests --global-setup setup.js
```
*setup.js*
```js
exports.default = {
    before: async t => {},
    after: async t => {}
    beforeEach: async t => {},
    afterEach: async t => {}
    requestHooks: []
    page: ...
};
```

The feature will be available in all supported interfaces: command-line, programmatic, and configuration file.
Please note that it's not final and can be changed.

**Comment by benmonro** [comment]
@miherlosev that's great news. Is there a PR I can follow to track this?

**Comment by miherlosev** [comment]
Hi @benmonro 

No PR is available. You can see all active PRs here: https://github.com/DevExpress/testcafe/pulls
We will unlikely work on this until we finalize the features from our [roadmap](https://devexpress.github.io/testcafe/roadmap/).

**Comment by llwt** [comment]
@miherlosev Would you be open to a PR to get this in sooner?

**Comment by miherlosev** [comment]
https://github.com/DevExpress/testcafe/issues/4323

@llwt 

I am working on finalizing the feature from [roadmap](https://devexpress.github.io/testcafe/roadmap/) and now I don't have time to implement this feature.

**Comment by llwt** [comment]
@miherlosev Sorry if I was unclear. Would you be open to someone from our team submitting a PR to implement this, with your guidance around what you'd like to see, since the previous PR was closed in favor of this approach?

**Comment by arubtsov** [comment]
@llwt, A PR is always welcomed.

**Comment by miherlosev** [comment]
https://stackoverflow.com/questions/65119877/how-to-add-custom-request-header-to-testcafe-test-suite
